### PR TITLE
Controllers and services for entities

### PR DESCRIPTION
**Pull request by bendgomin333**
Add controllers and services for entities, fixed the config file to be able to run migrations

### CODE REVIEW

File: src/entities/election/election.controller.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Controller, Get, Param, ParseIntPipe } from '@nestjs/common';
+import { ElectionService } from './election.service';
+
+@Controller('api/v1/elections')
```
Comment (shamilkhan) [inline src/entities/election/election.controller.ts]: Let's extract the prefix `api/v1` in highest scope.

File: src/config/typeorm.config.ts [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
   type: 'postgres',
   host: process.env.DB_HOST || 'localhost',
   port: parseInt(process.env.DB_PORT) || 5432,
   username: process.env.DB_USERNAME,
   password: process.env.DB_PASSWORD,
   database: process.env.DB_NAME || 'elections_us',
   entities: [__dirname + '/../**/*.entity{.ts,.js}'],
-  migrations: [__dirname + '/../migrations/{.ts,.js}'],
-  synchronize: true,
+  migrations: [__dirname + '/../migrations/*{.ts,.js}'],
```
Comment (shamilkhan) [inline src/config/typeorm.config.ts:11]: Let's add `migrationsTableName: 'migration'` to track all migrations

File: package.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
   "scripts": {
     "build": "nest build",
     "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
-    "start": "nest start",
+    "start": "typeorm-ts-node-esm migration:run -d src/config/typeorm.config.ts && nest start",
```
Comment (shamilkhan) [inline package.json]: Let's don't mix `start` with `migrations`. Could you, please, create commands like this (for all migrations cases): 
```
"scripts": {
  "typeorm": "ts-node ./node_modules/typeorm/cli",
  "typeorm:run-migrations": "npm run typeorm migration:run -- -d ./typeOrm.config.ts",
  "typeorm:generate-migration": "npm run typeorm -- -d ./typeOrm.config.ts migration:generate ./migrations/$npm_config_name",
  "typeorm:create-migration": "npm run typeorm -- migration:create ./migrations/$npm_config_name",
  "typeorm:revert-migration": "npm run typeorm -- -d ./typeOrm.config.ts migration:revert",
  ...
}
```

File: src/config/typeorm.config.ts [hunk 4]
```diff
@@ -1,15 +1,15 @@
-import { TypeOrmModuleOptions } from '@nestjs/typeorm';
+import { DataSource, DataSourceOptions } from 'typeorm';
 
-const typeOrmConfig: TypeOrmModuleOptions = {
+export const dataSourceOptions: DataSourceOptions = {
```
Comment (shamilkhan) [inline src/config/typeorm.config.ts:3]: Why did you change export and type?

File: src/app.module.ts [hunk 5]
```diff
@@ -1,11 +1,31 @@
...
+    StateModule,
+    CountyModule,
+    ElectionModule,
+    PartyModule,
+    CandidateModule,
+    CityElectionResultModule,
+    CountyElectionResultModule,
+    StateElectionResultModule,
+  ],
   controllers: [AppController],
```
Comment (shamilkhan) [inline src/app.module.ts]: Let's remove default Service and Controller

File: src/entities/candidate/candidate.controller.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { Controller, Get, Param, ParseIntPipe, Req } from '@nestjs/common';
```
Comment (shamilkhan) [inline src/entities/candidate/candidate.controller.ts]: Remove `Req`  import since it wasn't used in the controller methods.

File: src/entities/candidate/candidate.controller.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import { Candidate } from './candidate.entity';
+import { CandidateService } from './candidate.service';
+
+@Controller('api/v1/candidates')
+export class CandidateController {
+  constructor(private candidateService: CandidateService) {}
```
Comment (shamilkhan) [inline src/entities/candidate/candidate.controller.ts]: Please, Make candidateService property readonly in the constructor.

File: src/entities/candidate/candidate.service.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    @InjectRepository(Candidate)
+    private candidateRepository: Repository<Candidate>,
+  ) {}
+
+  findAll(): Promise<Candidate[]> {
+    return this.candidateRepository.find();
+  }
+
+  findOneById(id: number): Promise<Candidate | null> {
+    return this.candidateRepository.findOneBy({ id });
```
Comment (shamilkhan) [inline src/entities/candidate/candidate.service.ts:19]: Maybe it's better to use
`this.candidateRepository.findOne(id);` or`this.candidateService.findOneById(id);`?

File: src/entities/candidate/candidate.service.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { Candidate } from './candidate.entity';
+
+@Injectable()
+export class CandidateService {
+  constructor(
+    @InjectRepository(Candidate)
```
Comment (shamilkhan) [inline src/entities/candidate/candidate.service.ts:10]: Please, make candidateRepository property readonly in the constructor.

File: src/entities/state/state.service.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  findAll(): Promise<State[]> {
+    return this.stateRepository.find();
+  }
+
+  findOne(id: number): Promise<State> {
+    return this.stateRepository.findOneBy({ id });
+  }
+
+  findElectionResults(id: number): Promise<StateElectionResult[]> {
```
Comment (shamilkhan) [inline src/entities/state/state.service.ts]: Why do we need to duplicate this method in State if we have `state-election-entity` with results?

File: src/entities/party/party.service.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    @InjectRepository(Party)
+    private readonly partyRepository: Repository<Party>,
+  ) {}
+
+  findAll(): Promise<Party[]> {
+    return this.partyRepository.find();
+  }
+
+  findOne(id: number): Promise<Party> {
+    return this.partyRepository.findOneBy({ id });
```
Comment (shamilkhan) [inline src/entities/party/party.service.ts:19]: Let's use everywhere the standard way to search by id. I mean we use 2 versions `findOne(id)`, `findOneBy({ id })` let's use one of them.

File: src/entities/county/county.module.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import { Module } from '@nestjs/common';
+import { TypeOrmModule } from '@nestjs/typeorm';
+import { County } from './county.entity';
+import { CountyController } from './county.controller';
+import { CountyService } from './county.service';
+import { CountyElectionResultModule } from '../county-election-result/county-election-result.module';
+
+@Module({
+  imports: [TypeOrmModule.forFeature([County]), CountyElectionResultModule],
```
Comment (shamilkhan) [inline src/entities/county/county.module.ts:9]: Please, don't mix `CountyElectionResultModule` with `Couty`

**Review by shamilkhan — CHANGES_REQUESTED** [review]

File: src/config/typeorm.config.ts (hunk 4 above)
Comment (bendgomin333) [reply src/config/typeorm.config.ts:3]: The documentation for the latest typeorm version recommends to configure database connection in this way. Also, migrations work only if data-source file contain export of DataSource instance.
err

Comment (shamilkhan) [reply src/config/typeorm.config.ts:3]: Ok.

File: src/entities/election/election.controller.ts (hunk 1 above)
Comment (bendgomin333) [reply src/entities/election/election.controller.ts]: Done

File: src/config/typeorm.config.ts (hunk 2 above)
Comment (bendgomin333) [reply src/config/typeorm.config.ts:11]: By default `migrationTableName === 'migrations'`

File: src/entities/candidate/candidate.service.ts (hunk 9 above)
Comment (bendgomin333) [reply src/entities/candidate/candidate.service.ts:10]: Done, and fixed this for properties in another services and controllers

File: src/entities/candidate/candidate.controller.ts (hunk 6 above)
Comment (bendgomin333) [reply src/entities/candidate/candidate.controller.ts]: Done

File: src/app.module.ts (hunk 5 above)
Comment (bendgomin333) [reply src/app.module.ts]: Done

File: src/entities/candidate/candidate.controller.ts (hunk 7 above)
Comment (bendgomin333) [reply src/entities/candidate/candidate.controller.ts]: Done, also fixed same troubles in another controllers and services

File: src/entities/county/county.module.ts (hunk 12 above)
Comment (bendgomin333) [reply src/entities/county/county.module.ts:9]: Are you about imports on 6 and 9 lines? It's necessary to be able to inject `CountyElectionResultService` in `CountyService` constructor. 
For now I've implemented `/counties/:id/election-results` endpoint in the same controller (`CountyController`), near `/counties` and `/counties/:id`. Thus, all endpoints related to the county are located in one file.
If it's a bad decision and it generates unwanted dependencies, I can create controller for `CountyElectionResult` and put `/counties/:id/election-results` endpoint there.

Comment (shamilkhan) [reply src/entities/county/county.module.ts:9]: Ok. I've checked the API structure. I agree with you.

File: src/entities/candidate/candidate.service.ts (hunk 8 above)
Comment (bendgomin333) [reply src/entities/candidate/candidate.service.ts:19]: `findOneById` is deprecated, `findOne` looks like `findOneBy` but has more options for making a query (e.g. where/select/etc options)
`...repository.findOneBy({id}) `the same as` ...repository.findOne({ where: {id} })`

I renamed all functions in services and controllers in same style `findOneBy`

File: src/entities/party/party.service.ts (hunk 11 above)
Comment (bendgomin333) [reply src/entities/party/party.service.ts:19]: Done. I've selected second version (`findOneBy({id}`)

File: src/entities/state/state.service.ts (hunk 10 above)
Comment (bendgomin333) [reply src/entities/state/state.service.ts]: Done. I've removed duplicated methods in services and injected them into controllers

File: src/entities/candidate/candidate.service.ts (hunk 8 above)
Comment (shamilkhan) [reply src/entities/candidate/candidate.service.ts:19]: Thank you!
